### PR TITLE
image filtering node

### LINKED ISSUE #13: [TASK] Image filtering node
### Description of task

Create a node that filters a video stream. This is necessary for the ArUco detector #12 to work under water in rough conditions.

### Suggested Workflow

- [ ] Look at the [ROS1 node](https://github.com/vortexntnu/vortex-cv/tree/main/image_filters) I made earlier
- [ ] Do it in ROS2

### Specifications

- [ ] Config file/ROS params:
  - [ ] Change filtering algorithm and associated parameters
  - [ ] input topic name
  - [ ] output topic name

### Contacts

- @EirikKolas 

### Code Quality

- [ ] Every function in header files are documented (inputs/returns/exceptions)
- [ ] The project has automated tests that cover MOST of the functions and branches in functions (pytest/gtest)
- [ ] The code is documented on the wiki (provide link)

### PR DESCRIPTION
**Pull request by jorgenfj**

### CODE REVIEW

File: image_filters/params/image_filter_parameters.yaml [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+        mask_weight: 5
+
+
+#   NOFILTER = 0,
+#   SHARPENING = 1,
+#   UNSHARPENING = 2,
+#   ERODING = 3,
+#   DILATING = 4,
+#   WHITE_BALANCING = 5,
+#   EBUS = 6
```
Comment (EirikKolas) [inline image_filters/params/image_filter_parameters.yaml]: Forgot to remove this :)

File: image_filters/test/image_processing_test.cpp
Comment (EirikKolas) [inline image_filters/test/image_processing_test.cpp:1]: bruh

File: image_filters/test/CMakeLists.txt
Comment (EirikKolas) [inline image_filters/test/CMakeLists.txt:1]: no tests :(

**Review by EirikKolas — CHANGES_REQUESTED** [review]
Very nice and clean ROS code 😎. I have two suggestions (as we talked about):
1. Standardize the input to the filter functions. 
  - by either creating a class that stores the parameter struct as a member variable as well as all the functions
  - or making all the functions take the same parameter struct as an argument.
2. Make the `std::map` point to the function instead of an enum. The enum will become redundant in this case.

This will make it a lot easier to add new filters later: Add one function and update the map with one entry. Now you have to update the map, the enum, and the switch-case in addition to adding the function.

File: image_filters/README.md
Comment (EirikKolas) [inline image_filters/README.md:1]: Amazing! 🤩

**Review by EirikKolas — APPROVED** [review]
Good job! Really liked the documentation

**Comment by jorgenfj** [comment]
Not finished yet. There is a CMake error when building

File: image_filters/params/image_filter_parameters.yaml (hunk 1 above)
Comment (EirikKolas) [reply image_filters/params/image_filter_parameters.yaml]: bump

**Comment by EirikKolas** [comment]
You can access elements in the map as `filter_map["name_of_filter"]` btw. Was too small of a change to bother you to change it, but since you are already working on it... :)

File: image_filters/src/image_processing.cpp [hunk 2]
```diff
@@ -24,86 +41,64 @@ void unsharpeningFilter(const cv::Mat &original, cv::Mat &filtered, size_t blurS
...
-		break;
+void apply_filter(const cv::Mat &original, cv::Mat &filtered, const FilterParams& filter_params)
+{
+	// Find the filter in the filter_functions map
+    auto it = filter_functions.find(filter_params.filter_type);
+    if (it != filter_functions.end()) {
+		// Calls the filter function
+        it->second(original, filtered, filter_params);
+    } else {
+        original.copyTo(filtered);  // Default to no filter
```
Comment (EirikKolas) [inline image_filters/src/image_processing.cpp]: This can be simplified :)

**Pull request closed without merging**